### PR TITLE
Client Certificate Authentication (Using Backend HTTPS Support)

### PR DESCRIPTION
**Pull request by eshryane**
Implement support for client certificate authentication, using support for HTTPS in the backend.

Related to  #637

### CODE REVIEW

**Comment by eshryane** [comment]
Integration tests passed [3f224ab](https://github.com/RIPE-NCC/whois/pull/1343/commits/3f224ab21fb8fa72a9fd8f8963de4c1e28c60a16)

File: whois-api/src/test/java/net/ripe/db/whois/api/rest/WhoisClientCertificateAuthenticationTestIntegration.java [hunk 1]
```diff
@@ -1,34 +1,41 @@
...
-import static net.ripe.db.whois.update.keycert.X509CertificateUtil.getCertificateAsString;
 import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.containsString;
+import static org.hamcrest.Matchers.hasSize;
 import static org.hamcrest.Matchers.is;
+import static org.junit.jupiter.api.Assertions.fail;
 
 @Tag("IntegrationTest")
-public class WhoisClientCertificateAuthenticationTestIntegration extends AbstractIntegrationTest {
+public class WhoisClientCertificateAuthenticationTestIntegration extends AbstractClientCertificateIntegrationTest {
```
Comment (MiguelAHM) [inline whois-api/src/test/java/net/ripe/db/whois/api/rest/WhoisClientCertificateAuthenticationTestIntegration.java:38]: Rename by ClientCertificateAuthenticationTestIntegration?

File: whois-api/src/main/java/net/ripe/db/whois/api/UpdatesParser.java [hunk 2]
```diff
@@ -62,12 +63,15 @@ public List<Paragraph> createParagraphs(final ContentWithCredentials contentWith
             baseCredentials.add(SsoCredential.createOfferedCredential(updateContext.getUserSession()));
         }
 
-        updateContext.getClientCertificate().ifPresent(x509 -> baseCredentials.add(ClientCertificateCredential.createOfferedCredential(x509)));
+        if (updateContext.getClientCertificates() != null) {
```
Comment (MiguelAHM) [inline whois-api/src/main/java/net/ripe/db/whois/api/UpdatesParser.java:66]: Just a suggestion
I think that this code is duplicated in some other classes. Maybe we can put this code just in a common class and call it from the rest.

Same happen with this method:
 public void setClientCertificates(final UpdateContext updateContext, final HttpServletRequest request) {
    updateContext.setClientCertificates(ClientCertificateExtractor.getClientCertificates(request));
}

However, it is not too much duplicated code so feel free to ignore

**Review by MiguelAHM — APPROVED** [review]

File: whois-api/src/test/java/net/ripe/db/whois/api/rest/WhoisClientCertificateAuthenticationTestIntegration.java (hunk 1 above)
Comment (eshryane) [reply whois-api/src/test/java/net/ripe/db/whois/api/rest/WhoisClientCertificateAuthenticationTestIntegration.java]: Good idea, but I renamed it to `WhoisRestServiceClientCertificateTestIntegration` to be consistent with the other REST test classes (it contains REST API tests).

File: whois-api/src/main/java/net/ripe/db/whois/api/UpdatesParser.java (hunk 2 above)
Comment (eshryane) [reply whois-api/src/main/java/net/ripe/db/whois/api/UpdatesParser.java:66]: If you look at the 3 places this code is duplicated, there is not much overlap (it's used for different types of updates) so de-duplicating those lines are not straightforward. It's also the case for the other methods.

File: whois-api/src/test/java/net/ripe/db/whois/api/httpserver/CertificatePrivateKeyPair.java [hunk 3]
```diff
@@ -30,6 +33,8 @@ public class CertificatePrivateKeyPair {
         Security.addProvider(new BouncyCastleProvider());
     }
 
+    private X509Certificate certificate;
```
[the remark this answers is not in the document]
Comment (eshryane) [reply whois-api/src/test/java/net/ripe/db/whois/api/httpserver/CertificatePrivateKeyPair.java]: Fixed, thanks

**Comment by eshryane** [comment]
Integration tests passed [707279d](https://github.com/RIPE-NCC/whois/pull/1343/commits/707279d01fc8cc199eae3954f64e043288f721b8)

**Pull request merged**